internal/liveness/metrics.go

```diff
@@ -0,0 +1,135 @@
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package liveness
+
+import (
+	"context"
+	"strings"
+	"time"
+
+	"github.com/ceph/ceph-csi/internal/util"
+	"github.com/prometheus/client_golang/prometheus"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/klog"
+)
+
+var (
+	rbdMapping = *prometheus.NewGaugeVec(prometheus.GaugeOpts{
+		Namespace: "csi",
+		Name:      "ceph_rbd",
+		Help:      "ceph rbd image pvc mapping"},
+		[]string{"pvc_name", "pvc_namespace", "pv_name", "pool", "imageID"})
+	cephfsMapping = *prometheus.NewGaugeVec(prometheus.GaugeOpts{
+		Namespace: "csi",
+		Name:      "ceph_fs",
+		Help:      "cephfs subvolume pvc mapping"},
+		[]string{"pvc_name", "pvc_namespace", "pv_name", "volume_name", "subvolume_group", "subvolume_name"})
+)
+
+func getMapping(timeout time.Duration) {
+	klog.V(4).Infof("Start recording mapping")
+	client := util.NewK8sClient()
+	pvcs, err := client.CoreV1().PersistentVolumeClaims("").List(context.TODO(), metav1.ListOptions{})
+	if err != nil {
+		klog.Errorf("Get pvclist failed: %v", err)
+		return
+	}
+	for _, pvc := range pvcs.Items {
+		pvcName := pvc.Name
+		pvcNameSpace := pvc.Namespace
+		pvName := pvc.Spec.VolumeName
+		pvInfo, err := client.CoreV1().PersistentVolumes().Get(context.TODO(), pvName, metav1.GetOptions{})
+		if err != nil {
+			klog.V(4).Infof("Get pvinfo failed: %v", err)
+			continue
+		}
+		if pvInfo.Spec.CSI != nil {
+			volumeHandle := pvInfo.Spec.CSI.VolumeHandle
+			volumeHandleSlice := strings.Split(volumeHandle, "-")
+			index := len(volumeHandleSlice) - 5
+			imageName := strings.Join(append([]string{"csi-vol"}, volumeHandleSlice[index:]...), "-")
+			var secretName string
+			var secretNamespace string
+			if pvInfo.Spec.CSI.ControllerExpandSecretRef != nil {
+				secretName = pvInfo.Spec.CSI.ControllerExpandSecretRef.Name
+				secretNamespace = pvInfo.Spec.CSI.ControllerExpandSecretRef.Namespace
+			} else {
+				secretName = pvInfo.Spec.CSI.NodePublishSecretRef.Name
+				secretNamespace = pvInfo.Spec.CSI.NodePublishSecretRef.Namespace
+			}
+			imageCluster := pvInfo.Spec.CSI.VolumeAttributes["clusterID"]
+			monitors, err := util.Mons(csiConfigFile, imageCluster)
+			if err != nil {
+				klog.Error(err.Error())
+				return
+			}
+			secret, _ := client.CoreV1().Secrets(secretNamespace).Get(context.TODO(), secretName, metav1.GetOptions{})
+			newSecrets := make(map[string]string)
+			for key, value := range secret.Data {
+				newSecrets[key] = string(value)
+			}
+			cr, err := util.NewUserCredentials(newSecrets)
+			if err != nil {
+				klog.Error(err.Error())
+				return
+			}
+			imagePool := pvInfo.Spec.CSI.VolumeAttributes["pool"]
+			volumeName := pvInfo.Spec.CSI.VolumeAttributes["fsName"]
+			if imagePool == "" && volumeName == "" {
+				//unknown volume
```

Comment:
this should not happen, please log the pvc and pv name

Reply:
done